lib/phoenix/transports/long_poller.ex

```diff
@@ -79,75 +72,119 @@ defmodule Phoenix.Transports.LongPoller do
-    conn = put_session_with_salt(conn, server_pid)
+    priv_topic =
+      "phx:lp:"
+      |> Kernel.<>(Base.encode64(:crypto.strong_rand_bytes(16)))
+      |> Kernel.<>(:os.timestamp() |> Tuple.to_list |> Enum.join(""))
 
-    {conn, server_pid}
-  end
+    child = [router, timeout_window_ms(conn), priv_topic, pubsub_server(conn)]
+    {:ok, server_pid} = Supervisor.start_child(LongPoller.Supervisor, child)
 
-  # Serialized longpoll server pid into session and harden with random salt
-  defp put_session_with_salt(conn, server_pid) do
-    key = session_key(conn)
-    conn
-    |> put_session(key, :erlang.term_to_binary(server_pid))
-    |> put_session("#{key}_salt", :crypto.strong_rand_bytes(16) |> Base.encode64)
+    {conn, priv_topic, sign(conn, priv_topic), server_pid}
   end
 
   @doc """
   Finds the `Phoenix.LongPoller.Server` server from the session
   """
   def resume_session(conn) do
-    case longpoll_pid(conn) do
-      {:ok, pid}            -> {:ok, conn, pid}
-      :nopid                -> {:error, conn, :terminated}
+    case verify_longpoll_topic(conn) do
+      {:ok, priv_topic}     -> {:ok, conn, priv_topic}
+      :notopic              -> {:error, conn, :terminated}
       {:error, :terminated} -> {:error, conn, :terminated}
     end
   end
 
   @doc """
-  Retrieves the serialized `Phoenix.LongPoller.Server` pid from the session
+  Retrieves the serialized `Phoenix.LongPoller.Server` pid from the encrypted token
   """
-  def longpoll_pid(conn) do
-    case get_session(conn, session_key(conn)) do
-      nil -> :nopid
-      bin ->
-        pid = :erlang.binary_to_term(bin)
-        if Process.alive?(pid) do
-          {:ok, pid}
-        else
-          {:error, :terminated}
+  def verify_longpoll_topic(%Plug.Conn{params: %{"token" => token, "sig" => sig}} = conn) do
+    case verify(conn, token, sig) do
+      {:ok, priv_topic} ->
+        ref = :erlang.make_ref()
+        :ok = subscribe(conn, priv_topic)
+        :ok = broadcast_from(conn, priv_topic, {:subscribe, ref})
+        receive do
+          {:ok, :subscribe, ^ref} -> {:ok, priv_topic}
+        after
+          pubsub_timeout_ms(conn)  -> {:error, :terminated}
         end
+
+      _ -> :notopic
     end
   end
+  def verify_longpoll_topic(_conn), do: :notopic
 
   @doc """
   Ack's a list of `%Phoenix.Socket.Messages{}`'s back to the `Phoenix.LongPoller.Server`
 
   To be called after buffered messages have been relayed to client
   """
-  def ack(server_pid, messages) do
-    :ok = GenServer.call(server_pid, {:ack, messages})
-  end
-
-  defp session_key(conn), do: "#{router_module(conn)}_longpoll_pid"
-
-  @doc """
-  Sets the active listener process. Called by polling `conn`
-  """
-  def set_active_listener(server_pid, listener_pid) do
-    GenServer.call(server_pid, {:set_active_listener, listener_pid})
+  def ack(conn, priv_topic, messages) do
+    ref = :erlang.make_ref()
+    :ok = broadcast_from(conn, priv_topic, {:ack, messages, ref})
```

Comment:
The messages have no ordering but once they arrive in the long poller
server they are ordered in its buffer. Think about this. The server got
three messages: a, b and c. They are stored like this:

    [c, b, a]

Now the server got two more:

    [e, d, c, b, a]

At this moment, you dispatch them to the long polling connection. However,
they are still in the buffer, because there were not acked. Later on, you
got two more (and no ack yet):

    [g, f, e, d, c, b, a]

Now you got an ack. The ack contains five, the number of messages that you
can drop from the tail:

    [g, f]

It doesn't matter the order here because we are just acking the in-memory
messages. This will hold true as long as messages are sent one by one (i.e.
if pid1 hasn't ack, you cannot send messages to pid2).



*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Lead Developer*

On Thu, Feb 12, 2015 at 4:38 AM, Chris McCord <notifications@github.com>
wrote:

> In lib/phoenix/transports/long_poller.ex
> <https://github.com/phoenixframework/phoenix/pull/594#discussion_r24556416>
> :
>
> >
> >    @doc """
> >    Ack's a list of `%Phoenix.Socket.Messages{}`'s back to the `Phoenix.LongPoller.Server`
> >
> >    To be called after buffered messages have been relayed to client
> >    """
> > -  def ack(server_pid, messages) do
> > -    :ok = GenServer.call(server_pid, {:ack, messages})
> > -  end
> > -
> > -  defp session_key(conn), do: "#{router_module(conn)}_longpoll_pid"
> > -
> > -  @doc """
> > -  Sets the active listener process. Called by polling `conn`
> > -  """
> > -  def set_active_listener(server_pid, listener_pid) do
> > -    GenServer.call(server_pid, {:set_active_listener, listener_pid})
> > +  def ack(conn, priv_topic, messages) do
> > +    ref = :erlang.make_ref()
> > +    :ok = broadcast_from(conn, priv_topic, {:ack, messages, ref})
>
> I made the message buffer a list of {ref, %Message{}}'s and now I only ack
> the refs. I don't think we can guarantee ordering so just acking the length
> won't work. acking the refs avoids most of the copying.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phoenixframework/phoenix/pull/594/files#r24556416>.
>